[SPARK-13811][SPARK-13836] [SQL] Removed IsNotNull Constraints of Compound Expressions And Generated IsNotNull Constraints inside Not

gatorsmile opened the pull request:
#### What changes were proposed in this pull request?

We generate `IsNotNull` constraint for compound expressions. However, we should not add extra predicate conditions for these compound expressions. It will add extra costs for query processing. Normally, it will not filter out any row but requires extra processing. 

For example, 

``` SQL
testRelation.where('a + 'b === 1)
```

Without this PR, the plan will be like 

```
Filter (((a#0 + b#0) = 1) && isnotnull((a#0 + b#0)))
 +- LocalRelation [a#0,b#0,c#0]
```

After the PR, constraints-generated null filtering of compound expressions will not be added.

```
Filter ((a#0 + b#0) = 1))
 +- LocalRelation [a#0,b#0,c#0]
```

The solution is to remove the `IsNotNull` constraints for compound expressions. 

In addition, this PR can generate `IsNotNull` constraints for all the `BinaryComparison` inside `Not` except `EqualNullSafe`.

cc @sameeragarwal @nongli @marmbrus 
#### How was this patch tested?

Added a test case for Filter and another case for Join

gatorsmile committed:
no nulll filtering is added for compound expressions.

gatorsmile committed:
Merge remote-tracking branch 'upstream/master' into compoundNullFilter

SparkQA commented:
**[Test build #52907 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52907/consoleFull)** for PR 11649 at commit [`9cba8a3`](https://github.com/apache/spark/commit/9cba8a339a36dba83765954c26905b6d61926d67).

sameeragarwal commented:
@gatorsmile this change makes sense to me, however what are your thoughts on slightly widening the scope of this change and not generating `IsNotNull` constraints for any compound expression (in `constructIsNotNullConstraints`)? As you pointed out, I can't think of an optimization that can make use of these non-nullability constraints for compound expressions. Later, we can always extend `constructIsNotNullConstraints` to infer `isNotNull` constraints for individual attributes from compound expressions.

SparkQA commented:
**[Test build #52907 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52907/consoleFull)** for PR 11649 at commit [`9cba8a3`](https://github.com/apache/spark/commit/9cba8a339a36dba83765954c26905b6d61926d67).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52907/
Test PASSed.

gatorsmile commented:
@sameeragarwal True, we can add it back when needing it. Let me directly filter it out in `constructIsNotNullConstraints`

marmbrus commented:
I guess my question is why we are doing this for filters at all.  It makes sense for joins, but perhaps we should just eliminate that part of the rule.

gatorsmile commented:
@marmbrus 

Are you saying we should not add null filters for `Filter`? https://github.com/apache/spark/pull/11372 The original PR shows it can benefit some data sources like orc ? I am not very clear if this is the major motivation. cc @sameeragarwal 

In addition, we will still remove the `IsNotNull` filters for compound expressions from the `constructIsNotNullConstraints` anyway? Is my understanding right?

gatorsmile committed:
removed isnotnull of compound expressions from constraints.
added a support for operator Not for all the binary operators.

gatorsmile committed:
remove the space.

gatorsmile commented:
In this commit, three parts are done:
- removed the `IsNotNull` constraints for compound expressions
- added `IsNotNull` constraints support for all the `BinaryComparison` inside `Not` except `EqualNullSafe`.
- added two more test cases.

SparkQA commented:
**[Test build #52954 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52954/consoleFull)** for PR 11649 at commit [`baa2cda`](https://github.com/apache/spark/commit/baa2cda39134c68316491d22ee3a0c5bff26f52c).

sameeragarwal commented:
@marmbrus by adding additional `isNotNull` checks, we are hoping to generate code that can quickly short circuit the filter condition checks in the generated code: https://github.com/apache/spark/pull/11585

marmbrus commented:
I see.  We should probably comment on that in `NullFiltering` if that is the intention.  Anytime there are implicit contracts between components like this it makes things harder to reason about.

If we are going to make this general to BinaryComparisions then we should probably also comment there that subclasses are assumed to be null intolerant.

marmbrus commented:
Also, does codegen really need us to add redundant expressions into the query plan?  Can't it just look at the constraints and pick those that are useful for it?

gatorsmile commented:
Will add the comments to explain the null intolerant of the `BinaryComparisions` subclasses.

For avoiding the merge conflicts, will not touch the code in `NullFiltering`, since @sameeragarwal will rewrite the whole function for inferred predicate push down. @sameeragarwal Could you add a comment there? Thanks!

gatorsmile committed:
added a comment.

SparkQA commented:
**[Test build #52958 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52958/consoleFull)** for PR 11649 at commit [`55155a7`](https://github.com/apache/spark/commit/55155a76f9e7e01d85819352b3e0852d7b6f6a20).

SparkQA commented:
**[Test build #52954 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52954/consoleFull)** for PR 11649 at commit [`baa2cda`](https://github.com/apache/spark/commit/baa2cda39134c68316491d22ee3a0c5bff26f52c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52954/
Test FAILed.

sameeragarwal commented:
@marmbrus the reason for adding these inferred filters in the logical plan is because we currently only propagate constraints in the logical plan (and they are not directly accessible during code generation). 

In theory, we can propagate constraints in the physical plan instead but that'd be slightly unclean (since we'd then need to add propagation logic for all individual physical operators). @nongli, @yhuai and I briefly discussed the pros and cons of the 2 approaches yesterday and then decided to go with the current approach.

PS: Added comments in https://github.com/apache/spark/pull/11665.

SparkQA commented:
**[Test build #52958 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52958/consoleFull)** for PR 11649 at commit [`55155a7`](https://github.com/apache/spark/commit/55155a76f9e7e01d85819352b3e0852d7b6f6a20).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52958/
Test FAILed.

marmbrus commented:
I guess I would probably propagate them during planning (i.e. when you are creating the physical `Filter` operator) since its not really an optimization to add them to a logical plan, but instead its a physical concern that things execute more efficiently for that one specific operator if they are present.

The nice thing about this is then other operators that don't benefit from them (think `PhysicalOperation`) don't necessary need to see them.

gatorsmile commented:
I like the idea of @marmbrus 

Obviously, our outer join elimination is using null filtering constraints of compound expression. I think this PR has to wait until the other related PRs are merged.

sameeragarwal commented:
I do see the benefits of propagating constraints in the physical plan but if I understand correctly, wouldn't we have problems with joins?

For e.g., if a join has a condition on (`t1.a === t2.a && t1.a == 5`), just inferring the constraints/filters during physical planning wouldn't give us the liberty of pushing the additional `isNotNull(t1.a)`, `isNotNull(t2.a)` and `t2.a === 5` filters all the way to the leaf nodes on either side of the join in the query plan (there'd be no `PredicatePushdown` rules).

gatorsmile commented:
As mentioned in another thread, we do not explicitly add the inferred predicates(including `isNotNull`) into the conditions always. We just store them in the node. So far, we only do a bottom-up constraint propagation. That is not enough. We need a way to push down the local inferred predicates as low as possible. 

I am not sure what is the best way to implement it. I guess we need such a class parameter for storing constraints in each case class?

yhuai commented:
The benefit of adding those extra filters in the optimizer is that the predicate pushdown will just work.

marmbrus commented:
I see, the join case makes sense to me.  I like this solution.

sameeragarwal commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/ConstraintPropagationSuite.scala:
```diff
@@ -217,4 +217,19 @@ class ConstraintPropagationSuite extends SparkFunSuite {
         IsNotNull(resolveColumn(tr, "b")))))
   }
+
+  test("IsNotNull constraints of compound expressions in filters") {
```
nit: "no IsNotNull constraints are generated for compound expressions in filters" (here and in the test below)

sameeragarwal commented:
LGTM
